text/0000-extern-types.md

```diff
@@ -0,0 +1,128 @@
+
+In C it's possible to declare a type but not define it.
+These incomplete types can only be used behind pointers, a compilation error will result if the user tries to use them in such a way that the compiler would need to know their layout.
+
+In Rust, we don't have this feature. Instead, a couple of problematic hacks are used in its place.
+
+One is, we define the type as an uninhabited type. eg.
+
+```rust
+enum MyFfiType {}
+```
+
+Another is, we define the type with a private field and no methods to construct it.
+
+```rust
+struct MyFfiType {
+    _priv: (),
+}
+```
+
+The point of both these constructions is to prevent the user from being able to create or deal directly with instances of the type.
+Neither of these types accurately reflect the reality of the situation.
+The first definition is logically problematic as it defines a type which can never exist.
+This means that references to the type can also—logically—never exist and raw pointers to the type are guaranteed to be
+invalid.
+The second definition says that the type is a ZST, that we can store it on the stack and that we can call `ptr::read`, `mem::size_of` etc. on it.
+None of this is of course valid.
+
+The controversies on how to represent foreign types even extend to the standard library too; see the discussion in the [libc_types RFC PR](https://github.com/rust-lang/rfcs/pull/1783).
+
+This RFC instead proposes a way to directly express that a type exists but is unknown to Rust.
+
+Finally, In the 2017 roadmap, [integration with other languages](https://github.com/rust-lang/rfcs/blob/master/text/1774-roadmap-2017.md#integration-with-other-languages), is listed as a priority.
+Just like unions, this is an unsafe feature necessary for dealing with legacy code in a correct and understandable manner.
+
+# Detailed design
+[design]: #detailed-design
+
+Add a new kind of type declaration, an extern type:
+
+```rust
+extern {
+    type Foo;
+}
+```
+
+These types are FFI-safe. They are also DSTs, meaning that they do not implement `Sized`. Being DSTs, they cannot be kept on the stack and can only be accessed through pointers.
+
+In Rust, pointers to DSTs carry metadata about the object being pointed to.
+For strings and slices this is the length of the buffer, for trait objects this is the object's vtable.
+For extern types the metadata is simply `()`.
+This means that a pointer to an extern type is identical to a raw pointer.
+It also means that if we store an extern type at the end of a container (such as a struct or tuple) pointers to that container will also be identical to raw pointers (despite the container as a whole being unsized).
+This is useful to support a pattern found in some C APIs where structs are passed around which have arbitrary data appended to the end of them: eg.
+
+```rust
+extern {
+    type OpaqueTail;
+}
+
+#[repr(C)]
+struct FfiStruct {
+    data: u8,
+    more_data: u32,
+    tail: OpaqueTail,
```

Comment:
This structure seems worthless to me. Consider 3 cases where `FfiStruct` could be used:

```
fn by_return() -> FfiStruct;
fn by_sret(retptr: *mut FfiStruct);
fn by_argument(arg: FfiStruct);
```

Now, all of these are invalid or can’t be made work:

1. For `by_return` it may under covers be either a return by value or by sret pointer (analysed in next point); If its returned by value its all right. However you cannot really know if its by value or by sret without knowing the full defn' of structure;
2. For `by_sret` compiler simply cannot how much of stack space to allocate for the outptr slot;
3. For `by_argument` compiler cannot properly how to correctly pass such a structure to the C side (i.e. if the `FfiStruct` was supposed to be passed to C via registers, how many registers does the C side expect to be used?).

Basically, this thing is useless.